packages/typespec-python/generator/pygen/codegen/templates/model_dpg.py.jinja2

```diff
@@ -45,7 +45,8 @@
     __flattened_items = ["{{ model.flattened_items|join('\", \"') }}"]
     {% endif %}
 
-    {% if not model.internal and serializer.init_line(model) %}
+    {% set need_init = (not model.internal) and (serializer.init_line(model) or model.discriminator)%}
```

Comment:
For model which only has discriminator, we shall also generate `__init__` otherwise initialization of model will report error.